### PR TITLE
Transfer network bytes to smaller buffer

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently we read in 64KB blocks from the network. When TLS is not
enabled, these bytes are normally passed all the way to the application
layer (some exceptions: compression). For the HTTP layer this means that
these bytes can live throughout the entire lifecycle of an indexing
request.

The problem is that if the reads from the socket are small, this means
that 64KB buffers can be consumed by 1KB or smaller reads. If the socket
buffer or TCP buffer sizes are small, the leads to massive memory
waste. It has been identified as a major source of OOMs on coordinating
nodes as Elasticsearch easily exhausts the heap for these network bytes.

This commit resolves the problem by placing a handler after the TLS
handler to copy these bytes to a more appropriate buffer size as
necessary. This comes after TLS, because TLS is a framing layer which
often resolves this problem for us (the 64KB buffer will be decoded
into a more appropriate buffer size). However, this extra handler will
solve it for the non-TLS pipelines.

### CODE REVIEW

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyByteBufSizer.java
Comment (original-brownbear) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyByteBufSizer.java:37]: Cache `buf.maxFastWritableBytes()` and `buf.readableBytes()` to variables?

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyByteBufSizer.java
Comment (original-brownbear) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyByteBufSizer.java:39]: Might give us more optimally sized buffers if we set the max capacity as well through
via `ctx.alloc().heapBuffer(length, length);` ?

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyByteBufSizer.java
Comment (original-brownbear) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyByteBufSizer.java:33]: Should we really do this in general? It seems only makes sense for REST handlers that don't copy the buffers to unpooled anyway (search and bulk only as of right now). Maybe we should just copy those requests to new pooled buffers of appropriate size and leave the rest of them alone since we're releasing them on the io thread right away anyway?

**Review by original-brownbear — COMMENTED** [review]
Some smaller comments and a question.
Generally +1 on this, especially for searches we seem to be wasting a ton of bytes here.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyByteBufSizer.java
Comment (tbrooks8) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyByteBufSizer.java:33]: As discussed in the meeting, this is valuable as large messages can be aggregated for a period of time, hurting the memory ratios without this change.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyByteBufSizer.java
Comment (tbrooks8) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyByteBufSizer.java]: The max capacity does not impact the allocation size. It is essentially a limit for future expansion (reallocations) of the buffer.

**Comment by tbrooks8** [comment]
This is the simpler PR that I think we preferred.

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyByteBufSizer.java
Comment (original-brownbear) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyByteBufSizer.java:40]: Couldn't we, instead of rolling our own copying here, just use a call to `ByteBuf#capacity` and make Netty resize things? I.e. just do something like:

```
    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf buf, List<Object> out) {
        final int readableBytes = buf.readableBytes();
        out.add(buf.capacity(readableBytes).retain());
    }
```

At least in some quick experimentation with the debugger that method seems to accomplish a similar if not the same thing we do here now but maybe I'm missing something?

**Review by original-brownbear — COMMENTED** [review]
As discussed, I'm +1 on this solution over the alternative. Just one question.

**Review by henningandersen — APPROVED** [review]
LGTM, but let us have @original-brownbear's approval too before merging.

Could we maybe add a test just testing the logic in `NettyByteBufSizer` in isolation, verifying that we get all the bytes out in a reasonably sized buffer?

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyByteBufSizer.java
Comment (tbrooks8) [reply modules/transport-netty4/src/main/java/org/elasticsearch/transport/NettyByteBufSizer.java]: This seems fine. I guarded against less than 1024 because capacity will always copy small arrays which seems unnecessary.

**Comment by tbrooks8** [comment]
> Could we maybe add a test just testing the logic in NettyByteBufSizer in isolation, verifying that we get all the bytes out in a reasonably sized buffer?

I did but then removed it when I took Armin's suggestion. If we allow netty to control the resizing, it mutates the buffer in place and there is nothing to assert on. We depend that Netty internally reallocated and copied the bytes opposed to just adjusted indexes.

**Review by original-brownbear — APPROVED** [review]
LGTM, thanks Tim!

**Comment by tbrooks8** [comment]
Still needs back port to 7.9.x.

**Pull request merged**
